Issue starting Java Edge Module

- **OS and version used:** Ubuntu 16.04.5

# Description of the issue:
Everytime I try and start a default generated Edge Module on a Linux VM with IoT Edge running it fails. The default Edge modules for the other supported languages (Python, NodeJS) do work.

I could not find any earlier documentation on this issue.

# Console log of the issue:
```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
com.microsoft.azure.sdk.iot.device.exceptions.ModuleClientException: java.io.IOException: Broken pipe
        at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:252)
        at com.organisation.App.main(App.java:72)
Caused by: java.io.IOException: Broken pipe
        at jnr.enxio.channels.Common.write(Common.java:116)
        at jnr.enxio.channels.AbstractNativeSocketChannel.write(AbstractNativeSocketChannel.java:72)
        at jnr.unixsocket.UnixSocketChannel.write(UnixSocketChannel.java:261)
        at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendHttpRequestUsingUnixSocket(HttpsHsmClient.java:256)
        at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.sendRequestBasedOnScheme(HttpsHsmClient.java:225)
        at com.microsoft.azure.sdk.iot.device.hsm.HttpsHsmClient.getTrustBundle(HttpsHsmClient.java:157)
        at com.microsoft.azure.sdk.iot.device.edge.HttpsHsmTrustBundleProvider.getTrustBundleCerts(HttpsHsmTrustBundleProvider.java:38)
        at com.microsoft.azure.sdk.iot.device.ModuleClient.createFromEnvironment(ModuleClient.java:244)
        ... 1 more```

--- a/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java
+++ b/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceClient.java
@@ -393,6 +393,9 @@ public final class DeviceClient extends InternalClient implements Closeable
     /**
      * Asynchronously upload a stream to the IoT Hub.
      *
+     * NOTE: IotHub does not currently support CA signed devices using file upload. Please use SAS based authentication or
+     * self signed certificates.
+     *
      * @param destinationBlobName is a string with the name of the file in the storage.
      * @param inputStream is a InputStream with the stream to upload in the blob.
      * @param streamLength is a long with the number of bytes in the stream to upload.
@@ -402,10 +405,9 @@ public final class DeviceClient extends InternalClient implements Closeable
      * @throws IllegalArgumentException if the provided blob name, or the file path is {@code null},
      *          empty or not valid, or if the callback is {@code null}.
      * @throws IOException if the client cannot create a instance of the FileUpload or the transport.
-     * @throws UnsupportedOperationException if this method is called when using x509 authentication
      */
     public void uploadToBlobAsync(String destinationBlobName, InputStream inputStream, long streamLength,
-                                  IotHubEventCallback callback, Object callbackContext) throws IllegalArgumentException, IOException, UnsupportedOperationException
+                                  IotHubEventCallback callback, Object callbackContext) throws IllegalArgumentException, IOException
     {
         if (callback == null)
         {
@@ -423,11 +425,6 @@ public final class DeviceClient extends InternalClient implements Closeable
         }
 
         ParserUtility.validateBlobName(destinationBlobName);
-
-        if (this.config.getAuthenticationType() == DeviceClientConfig.AuthType.X509_CERTIFICATE)
-        {
-            throw new UnsupportedOperationException("File Upload does not support x509 authentication");
-        }
 
         if (this.fileUpload == null)
         {

--- a/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenAuthenticationProvider.java
+++ b/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenAuthenticationProvider.java
@@ -35,7 +35,7 @@ public abstract class IotHubSasTokenAuthenticationProvider extends IotHubAuthent
     protected IotHubSasToken sasToken;
 
     public abstract boolean canRefreshToken();
-    public abstract String getRenewedSasToken() throws IOException, TransportException;
+    public abstract String getRenewedSasToken(boolean proactivelyRenew) throws IOException, TransportException;
 
     public IotHubSasTokenAuthenticationProvider(String hostname, String gatewayHostname, String deviceId, String moduleId)
     {
@@ -71,28 +71,42 @@ public abstract class IotHubSasTokenAuthenticationProvider extends IotHubAuthent
         this.tokenValidSecs = tokenValidSecs;
     }
 
+    /**
+     * Returns true if the saved sas token needs to be manually renewed by the user
+     * @return true if the saved sas token needs to be manually renewed by the user
+     */
     public boolean isRenewalNecessary()
     {
         //Codes_SRS_IOTHUBSASTOKENAUTHENTICATION_34_017: [If the saved sas token has expired, this function shall return true.]
         return (this.sasToken != null && this.sasToken.isExpired());
     }
 
-    public boolean shouldRefreshToken()
+    /**
+     * Returns true if the saved token should be refreshed
+     *
+     * @param proactivelyRenew if true, this function will return true even if the saved token has not expired, but only
+     *                         if the token has lived beyond its time buffer percentage
+     * @return true if the caller should refresh the saved sas token
+     */
+    public boolean shouldRefreshToken(boolean proactivelyRenew)
     {
         if (this.sasToken.isExpired())
         {
             return true;
         }
 
-        long expiryTimeSeconds = IotHubSasToken.getExpiryTimeFromToken(this.sasToken.toString());
-        long tokenStartTime = expiryTimeSeconds - this.tokenValidSecs;
-        long bufferExpiryTime = this.tokenValidSecs * this.timeBufferPercentage / 100 + tokenStartTime;
-        long currentTimeSeconds = System.currentTimeMillis() / 1000;
-        if (bufferExpiryTime < currentTimeSeconds)
+        if (proactivelyRenew)
         {
-            //Codes_SRS_IOTHUBSASTOKENAUTHENTICATION_34_019: [This function shall return true if the saved token has lived for longer
-            // than its buffered threshold.]
-            return true;
+            long expiryTimeSeconds = IotHubSasToken.getExpiryTimeFromToken(this.sasToken.toString());
+            long tokenStartTime = expiryTimeSeconds - this.tokenValidSecs;
+            long bufferExpiryTime = this.tokenValidSecs * this.timeBufferPercentage / 100 + tokenStartTime;
+            long currentTimeSeconds = System.currentTimeMillis() / 1000;
+            if (bufferExpiryTime < currentTimeSeconds)
+            {
+                //Codes_SRS_IOTHUBSASTOKENAUTHENTICATION_34_019: [This function shall return true if the saved token has lived for longer
+                // than its buffered threshold.]
+                return true;
+            }
         }
 
         //Codes_SRS_IOTHUBSASTOKENAUTHENTICATION_34_020: [This function shall return false if the saved token has not lived for longer

--- a/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenHardwareAuthenticationProvider.java
+++ b/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenHardwareAuthenticationProvider.java
@@ -65,14 +65,19 @@ public class IotHubSasTokenHardwareAuthenticationProvider extends IotHubSasToken
     /**
      * Getter for SasToken. If the saved token has expired, this method shall renew it if possible
      *
+     * @param proactivelyRenew if true, this method will generate a fresh sas token even if the previously saved token
+     *                                 has not expired yet as long as the current token has lived beyond its buffer.
+     *                                 Use this for pre-emptively renewing sas tokens.
+
      * @throws IOException if generating the sas token from the TPM fails
      * @return The value of SasToken
      */
-    public String getRenewedSasToken() throws IOException
+    public String getRenewedSasToken(boolean proactivelyRenew) throws IOException
     {
-        if (this.sasToken.isExpired())
+        if (this.shouldRefreshToken(proactivelyRenew))
         {
             //Codes_SRS_IOTHUBSASTOKENHARDWAREAUTHENTICATION_34_035: [If the saved sas token has expired and there is a security provider, the saved sas token shall be refreshed with a new token from the security provider.]
+            //Codes_SRS_IOTHUBSASTOKENHARDWAREAUTHENTICATION_34_036: [If the saved sas token has not expired and there is a security provider, but the sas token should be proactively renewed, the saved sas token shall be refreshed with a new token from the security provider.]
             String sasTokenString = this.generateSasTokenSignatureFromSecurityProvider(this.tokenValidSecs);
             this.sasToken = new IotHubSasToken(this.hostname, this.deviceId, null, sasTokenString, this.moduleId, 0);
         }
@@ -122,6 +127,9 @@ public class IotHubSasTokenHardwareAuthenticationProvider extends IotHubSasToken
         throw new UnsupportedOperationException("Cannot change the trusted certificate when using security provider for authentication.");
     }
 
+    /**
+     * @return always returns false as the hardware authentication mechanism will never need to be updated with a new key or token
+     */
     @Override
     public boolean isRenewalNecessary()
     {

--- a/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenSoftwareAuthenticationProvider.java
+++ b/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenSoftwareAuthenticationProvider.java
@@ -97,16 +97,21 @@ public class IotHubSasTokenSoftwareAuthenticationProvider extends IotHubSasToken
     /**
      * Getter for SasToken. If the saved token has expired, this method shall renew it if possible
      *
+     * @param proactivelyRenew if true, this method will generate a fresh sas token even if the previously saved token
+     *                                 has not expired yet as long as the current token has lived beyond its buffer.
+     *                                 Use this for pre-emptively renewing sas tokens.
+     *
      * @return The value of SasToken
      */
     @Override
-    public String getRenewedSasToken() throws IOException, TransportException
+    public String getRenewedSasToken(boolean proactivelyRenew) throws IOException, TransportException
     {
-        if (this.sasToken.isExpired())
+        if (this.shouldRefreshToken(proactivelyRenew))
         {
             if (this.deviceKey != null)
             {
                 //Codes_SRS_IOTHUBSASTOKENSOFTWAREAUTHENTICATION_34_004: [If the saved sas token has expired and there is a device key present, the saved sas token shall be renewed.]
+                //Codes_SRS_IOTHUBSASTOKENAUTHENTICATION_34_006: [If the saved sas token has not expired and there is a device key present, but this method is called to proactively renew and the token should renew, the saved sas token shall be renewed.]
                 this.sasToken = new IotHubSasToken(this.hostname, this.deviceId, this.deviceKey, null, this.moduleId, getExpiryTimeInSeconds());
             }
         }

--- a/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenWithRefreshAuthenticationProvider.java
+++ b/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenWithRefreshAuthenticationProvider.java
@@ -79,9 +79,9 @@ public abstract class IotHubSasTokenWithRefreshAuthenticationProvider extends Io
      * @throws TransportException If a TransportException is encountered while refreshing the sas token
      */
     @Override
-    public String getRenewedSasToken() throws IOException, TransportException
+    public String getRenewedSasToken(boolean proactivelyRenew) throws IOException, TransportException
     {
-        if (this.shouldRefreshToken())
+        if (this.shouldRefreshToken(proactivelyRenew))
         {
             // Codes_SRS_MODULEAUTHENTICATIONWITHTOKENREFRESH_34_004: [This function shall invoke shouldRefreshSasToken, and if it should refresh, this function shall refresh the sas token.]
             this.refreshSasToken();

--- a/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/hsm/HttpsHsmClient.java
+++ b/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/hsm/HttpsHsmClient.java
@@ -17,6 +17,7 @@ import jnr.unixsocket.UnixSocketAddress;
 import jnr.unixsocket.UnixSocketChannel;
 
 import java.io.BufferedReader;
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.StringReader;
 import java.net.*;
@@ -249,11 +250,18 @@ public class HttpsHsmClient
             UnixSocketAddress address = new UnixSocketAddress(unixSocketAddress);
             channel = UnixSocketChannel.open(address);
 
-            channel.write(ByteBuffer.wrap(requestBytes));
-
             if (httpsRequest.getBody() != null)
             {
-                channel.write(ByteBuffer.wrap(httpsRequest.getBody()));
+                //append http request body to the request bytes
+                ByteArrayOutputStream outputStream = new ByteArrayOutputStream( );
+                outputStream.write(requestBytes);
+                outputStream.write(httpsRequest.getBody());
+
+                channel.write(ByteBuffer.wrap(outputStream.toByteArray()));
+            }
+            else
+            {
+                channel.write(ByteBuffer.wrap(requestBytes));
             }
 
             //read response

--- a/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsDeviceAuthenticationCBS.java
+++ b/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsDeviceAuthenticationCBS.java
@@ -302,7 +302,7 @@ public final class AmqpsDeviceAuthenticationCBS extends AmqpsDeviceAuthenticatio
         Section section = null;
         try
         {
-            section = new AmqpValue(deviceClientConfig.getSasTokenAuthentication().getRenewedSasToken());
+            section = new AmqpValue(deviceClientConfig.getSasTokenAuthentication().getRenewedSasToken(true));
             outgoingMessage.setBody(section);
         }
         catch (IOException e)

--- a/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/https/HttpsIotHubConnection.java
+++ b/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/https/HttpsIotHubConnection.java
@@ -445,7 +445,7 @@ public class HttpsIotHubConnection implements IotHubTransportConnection
     {
         try
         {
-            return this.config.getSasTokenAuthentication().getRenewedSasToken();
+            return this.config.getSasTokenAuthentication().getRenewedSasToken(false);
         }
         catch (IOException e)
         {

--- a/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java
+++ b/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java
@@ -123,7 +123,7 @@ public class MqttIotHubConnection implements IotHubTransportConnection, MqttMess
                 SSLContext sslContext = this.config.getAuthenticationProvider().getSSLContext();
                 if (this.config.getAuthenticationType() == DeviceClientConfig.AuthType.SAS_TOKEN)
                 {
-                    this.iotHubUserPassword = this.config.getSasTokenAuthentication().getRenewedSasToken();
+                    this.iotHubUserPassword = this.config.getSasTokenAuthentication().getRenewedSasToken(false);
                 }
                 else if (this.config.getAuthenticationType() == DeviceClientConfig.AuthType.X509_CERTIFICATE)
                 {

--- a/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java
+++ b/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/DeviceClientTest.java
@@ -1854,32 +1854,6 @@ public class DeviceClientTest
         Deencapsulation.invoke(client, "uploadToBlobAsync", destinationBlobName, mockInputStream, streamLength, mockedStatusCB, mockedPropertyCB);
     }
 
-    // Tests_SRS_INTERNALCLIENT_34_066: [If this function is called when the device client is using x509 authentication, an UnsupportedOperationException shall be thrown.]
-    @Test (expected = UnsupportedOperationException.class)
-    public void startFileUploadUploadToBlobAsyncAuthTypeThrows(@Mocked final FileUpload mockedFileUpload,
-                                                               @Mocked final InputStream mockInputStream,
-                                                               @Mocked final IotHubEventCallback mockedStatusCB,
-                                                               @Mocked final PropertyCallBack mockedPropertyCB) throws IOException, URISyntaxException, TransportException
-    {
-        //arrange
-        final IotHubClientProtocol protocol = IotHubClientProtocol.AMQPS;
-        final String destinationBlobName = "valid/blob/name.txt";
-        final long streamLength = 100;
-
-        new NonStrictExpectations()
-        {
-            {
-                mockConfig.getAuthenticationType();
-                result = DeviceClientConfig.AuthType.X509_CERTIFICATE;
-            }
-        };
-        DeviceClient client = Deencapsulation.newInstance(DeviceClient.class, new Class[] {String.class, IotHubClientProtocol.class}, "some conn string", protocol);
-        Deencapsulation.setField(client, "config", mockConfig);
-
-        // act
-        Deencapsulation.invoke(client, "uploadToBlobAsync", destinationBlobName, mockInputStream, streamLength, mockedStatusCB, mockedPropertyCB);
-    }
-
     /* Tests_SRS_INTERNALCLIENT_21_051: [If uploadToBlobAsync failed to start the upload using the FileUpload, it shall bypass the exception.] */
     @Test (expected = IllegalArgumentException.class)
     public void startFileUploadUploadToBlobAsyncThrows(@Mocked final FileUpload mockedFileUpload,

--- a/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenAuthenticationProviderTest.java
+++ b/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenAuthenticationProviderTest.java
@@ -78,7 +78,7 @@ public class IotHubSasTokenAuthenticationProviderTest
         }
 
         @Override
-        public String getRenewedSasToken() throws IOException
+        public String getRenewedSasToken(boolean proactivelyRenew) throws IOException
         {
             return null;
         }
@@ -247,7 +247,7 @@ public class IotHubSasTokenAuthenticationProviderTest
         IotHubSasTokenAuthenticationProvider authenticationProvider = new mockIotHubSasTokenAuthenticationImplementation(10, 1);
 
         //act
-        boolean result = authenticationProvider.shouldRefreshToken();
+        boolean result = authenticationProvider.shouldRefreshToken(true);
 
         //assert
         assertTrue(result);
@@ -262,7 +262,7 @@ public class IotHubSasTokenAuthenticationProviderTest
         IotHubSasTokenAuthenticationProvider authenticationProvider = new mockIotHubSasTokenAuthenticationImplementation(100000, 100);
 
         //act
-        boolean result = authenticationProvider.shouldRefreshToken();
+        boolean result = authenticationProvider.shouldRefreshToken(true);
 
         //assert
         assertFalse(result);

--- a/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenHardwareAuthenticationProviderTest.java
+++ b/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenHardwareAuthenticationProviderTest.java
@@ -136,7 +136,7 @@ public class IotHubSasTokenHardwareAuthenticationProviderTest
 
     //Tests_SRS_IOTHUBSASTOKENHARDWAREAUTHENTICATION_34_035: [If the saved sas token has expired and there is a security provider, the saved sas token shall be refreshed with a new token from the security provider.]
     @Test
-    public void getRenewedSasTokenAutoRenewsFromSecurityProvider(@Mocked final System systemMock) throws CertificateException, NoSuchAlgorithmException, KeyStoreException, KeyManagementException, IOException, SecurityProviderException, InvalidKeyException
+    public void getRenewedSasTokenAutoRenewsFromSecurityProvider() throws CertificateException, NoSuchAlgorithmException, KeyStoreException, KeyManagementException, IOException, SecurityProviderException, InvalidKeyException
     {
         //arrange
         final String someToken = "someToken";
@@ -170,7 +170,46 @@ public class IotHubSasTokenHardwareAuthenticationProviderTest
         IotHubSasTokenAuthenticationProvider sasAuth = new IotHubSasTokenHardwareAuthenticationProvider(expectedHostname, expectedGatewayHostname, expectedDeviceId, expectedModuleId, mockSecurityProviderTpm);
 
         //act
-        Deencapsulation.invoke(sasAuth, "getRenewedSasToken");
+        Deencapsulation.invoke(sasAuth, "getRenewedSasToken", false);
+    }
+
+    //Tests_SRS_IOTHUBSASTOKENHARDWAREAUTHENTICATION_34_036: [If the saved sas token has not expired and there is a security provider, but the sas token should be proactively renewed, the saved sas token shall be refreshed with a new token from the security provider.]
+    @Test
+    public void getRenewedSasTokenProactivelyRenewsFromSecurityProvider() throws CertificateException, NoSuchAlgorithmException, KeyStoreException, KeyManagementException, IOException, SecurityProviderException, InvalidKeyException
+    {
+        //arrange
+        final String someToken = "someToken";
+        final byte[] tokenBytes= someToken.getBytes();
+
+        //assert
+        new Expectations()
+        {
+            {
+                URLEncoder.encode(anyString, encodingName);
+                result = someToken;
+
+                System.currentTimeMillis();
+                result = 0;
+
+                mockSecurityProviderTpm.signWithIdentity((byte[]) any);
+                result = tokenBytes;
+                times = 2;
+
+                Base64.encodeBase64Local((byte[]) any);
+                result = tokenBytes;
+
+                URLEncoder.encode(anyString, encodingName);
+                result = someToken;
+
+                Deencapsulation.invoke(mockSasToken, "isExpired");
+                result = false;
+            }
+        };
+
+        IotHubSasTokenAuthenticationProvider sasAuth = new IotHubSasTokenHardwareAuthenticationProvider(expectedHostname, expectedGatewayHostname, expectedDeviceId, expectedModuleId, mockSecurityProviderTpm);
+
+        //act
+        Deencapsulation.invoke(sasAuth, "getRenewedSasToken", true);
     }
 
     //Tests_SRS_IOTHUBSASTOKENHARDWAREAUTHENTICATION_34_005: [This function shall return the saved sas token.]
@@ -201,7 +240,7 @@ public class IotHubSasTokenHardwareAuthenticationProviderTest
         IotHubSasTokenAuthenticationProvider sasAuth = new IotHubSasTokenHardwareAuthenticationProvider(expectedHostname, expectedGatewayHostname, expectedDeviceId, expectedModuleId, mockSecurityProviderTpm);
 
         //act
-        String actualSasToken = Deencapsulation.invoke(sasAuth, "getRenewedSasToken");
+        String actualSasToken = Deencapsulation.invoke(sasAuth, "getRenewedSasToken", false);
 
         //assert
         assertEquals(mockSasToken.toString(), actualSasToken);

--- a/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenSoftwareIotHubAuthenticationProviderTest.java
+++ b/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenSoftwareIotHubAuthenticationProviderTest.java
@@ -133,8 +133,70 @@ public class IotHubSasTokenSoftwareIotHubAuthenticationProviderTest
         IotHubSasTokenAuthenticationProvider sasAuth = new IotHubSasTokenSoftwareAuthenticationProvider(expectedHostname, expectedGatewayHostname, expectedDeviceId, expectedModuleId, expectedDeviceKey, expectedSasToken);
 
         //act
-        sasAuth.getRenewedSasToken();
+        sasAuth.getRenewedSasToken(false);
 
+    }
+
+    //Tests_SRS_IOTHUBSASTOKENAUTHENTICATION_34_006: [If the saved sas token has not expired and there is a device key present, but this method is called to proactively renew and the token should renew, the saved sas token shall be renewed.]
+    @Test
+    public void getRenewedSasTokenProactivelyRenews(@Mocked final System mockSystem) throws IOException, TransportException
+    {
+        new MockUp<IotHubSasTokenAuthenticationProvider>()
+        {
+            @Mock public boolean shouldRefreshToken(boolean proactivelyRenew)
+            {
+                return true;
+            }
+        };
+
+        //assert
+        new Expectations()
+        {
+            {
+                System.currentTimeMillis();
+                result = 0;
+                System.currentTimeMillis();
+                result = 0;
+                Deencapsulation.newInstance(IotHubSasToken.class, new Class[] {String.class, String.class, String.class, String.class, String.class, long.class}, expectedHostname, expectedDeviceId, expectedDeviceKey, null, expectedModuleId, expectedExpiryTime);
+                result = mockSasToken;
+            }
+        };
+
+        //arrange
+        IotHubSasTokenAuthenticationProvider sasAuth = new IotHubSasTokenSoftwareAuthenticationProvider(expectedHostname, expectedGatewayHostname, expectedDeviceId, expectedModuleId, expectedDeviceKey, expectedSasToken);
+
+        //act
+        sasAuth.getRenewedSasToken(true);
+    }
+
+    //Tests_SRS_IOTHUBSASTOKENAUTHENTICATION_34_006: [If the saved sas token has not expired and there is a device key present, but this method is called to proactively renew and the token should renew, the saved sas token shall be renewed.]
+    @Test
+    public void getRenewedSasTokenDoesntProactivelyRenewIfShouldntRefreshToken(@Mocked final System mockSystem) throws IOException, TransportException
+    {
+        new MockUp<IotHubSasTokenAuthenticationProvider>()
+        {
+            @Mock public boolean shouldRefreshToken(boolean proactivelyRenew)
+            {
+                return false;
+            }
+        };
+
+        //assert
+        new Expectations()
+        {
+            {
+                System.currentTimeMillis();
+                result = 0;
+                System.currentTimeMillis();
+                result = 0;
+            }
+        };
+
+        //arrange
+        IotHubSasTokenAuthenticationProvider sasAuth = new IotHubSasTokenSoftwareAuthenticationProvider(expectedHostname, expectedGatewayHostname, expectedDeviceId, expectedModuleId, expectedDeviceKey, expectedSasToken);
+
+        //act
+        sasAuth.getRenewedSasToken(true);
     }
 
     //Tests_SRS_IOTHUBSASTOKENSOFTWAREAUTHENTICATION_34_005: [This function shall return the saved sas token.]
@@ -157,7 +219,7 @@ public class IotHubSasTokenSoftwareIotHubAuthenticationProviderTest
         IotHubSasTokenAuthenticationProvider sasAuth = new IotHubSasTokenSoftwareAuthenticationProvider(expectedHostname, expectedGatewayHostname, expectedDeviceId, expectedModuleId, expectedDeviceKey, expectedSasToken);
 
         //act
-        String actualSasToken = Deencapsulation.invoke(sasAuth, "getRenewedSasToken");
+        String actualSasToken = Deencapsulation.invoke(sasAuth, "getRenewedSasToken", false);
 
         //assert
         assertEquals(mockSasToken.toString(), actualSasToken);

--- a/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenWithRefreshAuthenticationProviderTest.java
+++ b/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenWithRefreshAuthenticationProviderTest.java
@@ -38,7 +38,7 @@ public class IotHubSasTokenWithRefreshAuthenticationProviderTest
         }
 
         @Override
-        public boolean shouldRefreshToken()
+        public boolean shouldRefreshToken(boolean proactivelyRenew)
         {
             return this.shouldRefresh;
         }
@@ -126,7 +126,7 @@ public class IotHubSasTokenWithRefreshAuthenticationProviderTest
         moduleAuthenticationWithTokenRefresh.nextToken = newSasToken;
 
         //act
-        String actual = moduleAuthenticationWithTokenRefresh.getRenewedSasToken();
+        String actual = moduleAuthenticationWithTokenRefresh.getRenewedSasToken(true);
 
         //assert
         assertEquals(newSasToken.toString(), actual);
@@ -146,10 +146,9 @@ public class IotHubSasTokenWithRefreshAuthenticationProviderTest
         moduleAuthenticationWithTokenRefresh.nextToken = newSasToken;
 
         //act
-        String actual = moduleAuthenticationWithTokenRefresh.getRenewedSasToken();
+        String actual = moduleAuthenticationWithTokenRefresh.getRenewedSasToken(false);
 
         //assert
         assertEquals(oldSasToken.toString(), actual);
     }
-
 }

--- a/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/hsm/HttpsHsmClientTest.java
+++ b/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/hsm/HttpsHsmClientTest.java
@@ -262,6 +262,9 @@ public class HttpsHsmClientTest
                 HttpsRequestResponseSerializer.deserializeResponse((BufferedReader) any);
                 times = 1;
 
+                mockedUnixSocketChannel.write((ByteBuffer) any);
+                times = 1;
+
                 mockedHttpsRequest.send();
                 times = 0;
             }

--- a/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/https/HttpsIotHubConnectionTest.java
+++ b/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/https/HttpsIotHubConnectionTest.java
@@ -66,7 +66,7 @@ public class HttpsIotHubConnectionTest
         new NonStrictExpectations()
         {
             {
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result=testSasToken;
             }
         };
@@ -262,7 +262,7 @@ public class HttpsIotHubConnectionTest
                 result = iotHubHostname;
                 mockConfig.getDeviceId();
                 result = deviceId;
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result = tokenStr;
                 mockConfig.getAuthenticationType();
                 result = DeviceClientConfig.AuthType.SAS_TOKEN;
@@ -627,7 +627,7 @@ public class HttpsIotHubConnectionTest
                 result = iotHubHostname;
                 mockConfig.getDeviceId();
                 result = deviceId;
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result = tokenStr;
             }
         };
@@ -876,7 +876,7 @@ public class HttpsIotHubConnectionTest
                 result = iotHubHostname;
                 mockConfig.getDeviceId();
                 result = deviceId;
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result = tokenStr;
             }
         };
@@ -1517,7 +1517,7 @@ public class HttpsIotHubConnectionTest
                 result = iotHubHostname;
                 mockConfig.getDeviceId();
                 result = deviceId;
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result = tokenStr;
             }
         };
@@ -1706,7 +1706,7 @@ public class HttpsIotHubConnectionTest
         new Verifications()
         {
             {
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 times = 1;
             }
         };
@@ -1733,7 +1733,7 @@ public class HttpsIotHubConnectionTest
         new Verifications()
         {
             {
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 times = 1;
             }
         };

--- a/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnectionTest.java
+++ b/device/iot-device-client/src/test/java/tests/unit/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnectionTest.java
@@ -270,7 +270,7 @@ public class MqttIotHubConnectionTest
             {
                 mockConfig.getAuthenticationType();
                 result = DeviceClientConfig.AuthType.SAS_TOKEN;
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result = expectedSasToken;
                 mockConfig.isUseWebsocket();
                 result = false;
@@ -320,7 +320,7 @@ public class MqttIotHubConnectionTest
             {
                 mockConfig.getAuthenticationType();
                 result = DeviceClientConfig.AuthType.SAS_TOKEN;
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result = expectedToken;
                 mockConfig.isUseWebsocket();
                 result = true;
@@ -365,7 +365,7 @@ public class MqttIotHubConnectionTest
             {
                 mockConfig.getAuthenticationType();
                 result = DeviceClientConfig.AuthType.SAS_TOKEN;
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result = expectedToken;
                 mockConfig.isUseWebsocket();
                 result = false;
@@ -433,7 +433,7 @@ public class MqttIotHubConnectionTest
             {
                 mockConfig.getAuthenticationType();
                 result = DeviceClientConfig.AuthType.SAS_TOKEN;
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result = expectedToken;
                 new MqttMessaging(mockedMqttConnection, anyString, (IotHubListener) any, null, null, anyString, anyBoolean);
                 result = mockDeviceMessaging;
@@ -489,7 +489,7 @@ public class MqttIotHubConnectionTest
             {
                 mockConfig.getAuthenticationType();
                 result = DeviceClientConfig.AuthType.SAS_TOKEN;
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result = expectedToken;
                 new MqttMessaging(mockedMqttConnection, anyString, (IotHubListener) any, null, null, anyString, anyBoolean);
                 result = mockDeviceMessaging;
@@ -1070,7 +1070,7 @@ public class MqttIotHubConnectionTest
             {
                 mockConfig.getAuthenticationType();
                 result = DeviceClientConfig.AuthType.SAS_TOKEN;
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result = expectedSasToken;
                 mockConfig.isUseWebsocket();
                 result = false;
@@ -1107,7 +1107,7 @@ public class MqttIotHubConnectionTest
             {
                 mockConfig.getAuthenticationType();
                 result = DeviceClientConfig.AuthType.SAS_TOKEN;
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result = expectedSasToken;
                 mockConfig.isUseWebsocket();
                 result = false;
@@ -1155,7 +1155,7 @@ public class MqttIotHubConnectionTest
             {
                 mockConfig.getAuthenticationType();
                 result = DeviceClientConfig.AuthType.SAS_TOKEN;
-                mockConfig.getSasTokenAuthentication().getRenewedSasToken();
+                mockConfig.getSasTokenAuthentication().getRenewedSasToken(false);
                 result = expectedSasToken;
                 mockConfig.isUseWebsocket();
                 result = false;

--- a/iot-e2e-tests/android/app/src/androidTest/java/com/microsoft/azure/sdk/iot/android/iothubservices/FileUploadAndroidRunner.java
+++ b/iot-e2e-tests/android/app/src/androidTest/java/com/microsoft/azure/sdk/iot/android/iothubservices/FileUploadAndroidRunner.java
@@ -10,6 +10,8 @@ import com.microsoft.appcenter.espresso.ReportHelper;
 import com.microsoft.azure.sdk.iot.android.BuildConfig;
 import com.microsoft.azure.sdk.iot.android.helper.Rerun;
 import com.microsoft.azure.sdk.iot.common.tests.iothubservices.FileUploadTests;
+import com.microsoft.azure.sdk.iot.deps.util.Base64;
+
 import org.junit.BeforeClass;
 import org.junit.Ignore;
 import org.junit.Rule;
@@ -28,8 +30,13 @@ public class FileUploadAndroidRunner extends FileUploadTests
     @BeforeClass
     public static void setup() throws IOException
     {
+        String privateKeyBase64Encoded = BuildConfig.IotHubPrivateKeyBase64;
+        String publicKeyCertBase64Encoded = BuildConfig.IotHubPublicCertBase64;
         iotHubConnectionString = BuildConfig.IotHubConnectionString;
-        FileUploadTests.setUp();
+        String x509Thumbprint = BuildConfig.IotHubThumbprint;
+        String privateKey = new String(Base64.decodeBase64Local(privateKeyBase64Encoded.getBytes()));
+        String publicKeyCert = new String(Base64.decodeBase64Local(publicKeyCertBase64Encoded.getBytes()));
+        FileUploadTests.setUp(publicKeyCert, privateKey, x509Thumbprint);
     }
 
     @Ignore

--- a/iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/FileUploadTests.java
+++ b/iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/FileUploadTests.java
@@ -13,12 +13,10 @@ import com.microsoft.azure.sdk.iot.device.IotHubClientProtocol;
 import com.microsoft.azure.sdk.iot.device.IotHubEventCallback;
 import com.microsoft.azure.sdk.iot.device.IotHubStatusCode;
 import com.microsoft.azure.sdk.iot.service.*;
+import com.microsoft.azure.sdk.iot.service.auth.AuthenticationType;
 import com.microsoft.azure.sdk.iot.service.exceptions.IotHubException;
 import com.microsoft.azure.sdk.iot.service.exceptions.IotHubNotFoundException;
-import org.junit.After;
-import org.junit.AfterClass;
-import org.junit.Before;
-import org.junit.Test;
+import org.junit.*;
 
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
@@ -66,9 +64,14 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
     private static DeviceClient deviceClient;
     private static ServiceClient serviceClient;
     private static Device scDevice;
+    private static Device scDevicex509;
     private static FileUploadState[] fileUploadState;
     private static MessageState[] messageStates;
     private static FileUploadNotificationReceiver fileUploadNotificationReceiver;
+
+    private static String publicKeyCertificate;
+    private static String privateKeyCertificate;
+    private static String x509Thumbprint;
 
     enum STATUS
     {
@@ -127,11 +130,15 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
         }
     }
 
-    public static void setUp() throws IOException
+    public static void setUp(String publicK, String privateK, String thumbprint) throws IOException
     {
         registryManager = RegistryManager.createFromConnectionString(iotHubConnectionString);
         serviceClient = ServiceClient.createFromConnectionString(iotHubConnectionString, IotHubServiceClientProtocol.AMQPS);
 
+        publicKeyCertificate = publicK;
+        privateKeyCertificate = privateK;
+        x509Thumbprint = thumbprint;
+        
         fileUploadNotificationReceiver = serviceClient.getFileUploadNotificationReceiver();
         assertNotNull(fileUploadNotificationReceiver);
     }
@@ -142,6 +149,11 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
         String deviceId = "java-file-upload-e2e-test".concat(UUID.randomUUID().toString());
         scDevice = com.microsoft.azure.sdk.iot.service.Device.createFromId(deviceId, null, null);
         scDevice = registryManager.addDevice(scDevice);
+
+        String deviceIdX509 = "java-file-upload-e2e-test-x509".concat(UUID.randomUUID().toString());
+        scDevicex509 = com.microsoft.azure.sdk.iot.service.Device.createDevice(deviceIdX509, AuthenticationType.SELF_SIGNED);
+        scDevicex509.setThumbprint(x509Thumbprint, x509Thumbprint);
+        scDevicex509 = registryManager.addDevice(scDevicex509);
 
         // flush pending notifications before every test to prevent random test failures
         // because of notifications received from other failed test
@@ -177,6 +189,7 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
         try
         {
             registryManager.removeDevice(scDevice);
+            registryManager.removeDevice(scDevicex509);
         }
         catch (IotHubNotFoundException | IOException e)
         {
@@ -209,6 +222,12 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
     private void setUpDeviceClient(IotHubClientProtocol protocol) throws URISyntaxException, InterruptedException
     {
         deviceClient = new DeviceClient(DeviceConnectionString.get(iotHubConnectionString, scDevice), protocol);
+        IotHubServicesCommon.openClientWithRetry(deviceClient);
+    }
+
+    private void setUpX509DeviceClient(IotHubClientProtocol protocol) throws URISyntaxException, InterruptedException
+    {
+        deviceClient = new DeviceClient(DeviceConnectionString.get(iotHubConnectionString, scDevicex509), protocol, publicKeyCertificate, false, privateKeyCertificate, false);
         IotHubServicesCommon.openClientWithRetry(deviceClient);
     }
 
@@ -247,7 +266,7 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
         // act
         deviceClient.uploadToBlobAsync(fileUploadState[0].blobName, fileUploadState[0].fileInputStream, fileUploadState[0].fileLength, new FileUploadCallback(), fileUploadState[0]);
 
-        FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice();
+        FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice(scDevice);
 
         // assert
         verifyNotification(fileUploadNotification, fileUploadState[0]);
@@ -264,7 +283,7 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
 
         // act
         deviceClient.uploadToBlobAsync(fileUploadState[MAX_FILES_TO_UPLOAD - 1].blobName, fileUploadState[MAX_FILES_TO_UPLOAD - 1].fileInputStream, fileUploadState[MAX_FILES_TO_UPLOAD - 1].fileLength, new FileUploadCallback(), fileUploadState[MAX_FILES_TO_UPLOAD - 1]);
-        FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice();
+        FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice(scDevice);
 
         // assert
         assertNotNull(fileUploadNotification);
@@ -285,7 +304,7 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
         for (int i = 1; i < MAX_FILES_TO_UPLOAD; i++)
         {
             deviceClient.uploadToBlobAsync(fileUploadState[i].blobName, fileUploadState[i].fileInputStream, fileUploadState[i].fileLength, new FileUploadCallback(), fileUploadState[i]);
-            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice();
+            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice(scDevice);
 
             // assert
             verifyNotification(fileUploadNotification, fileUploadState[i]);
@@ -328,7 +347,7 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
                 }
             });
 
-            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice();
+            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice(scDevice);
 
             // assert
             verifyNotification(fileUploadNotification, fileUploadState[i]);
@@ -386,7 +405,7 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
                 }
             });
 
-            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice();
+            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice(scDevice);
 
             // assert
             verifyNotification(fileUploadNotification, fileUploadState[i]);
@@ -410,22 +429,68 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
         tearDownDeviceClient();
     }
 
-    private FileUploadNotification getFileUploadNotificationForThisDevice() throws IOException, InterruptedException
+    @Test (timeout = MAX_MILLISECS_TIMEOUT_KILL_TEST)
+    public void uploadToBlobAsyncAndTelemetryOnMQTTWS() throws URISyntaxException, IOException, InterruptedException, ExecutionException, TimeoutException
     {
-        FileUploadNotification fileUploadNotification;
-        do
+        // arrange
+        setUpDeviceClient(IotHubClientProtocol.MQTT_WS);
+        ExecutorService executor = Executors.newFixedThreadPool(2);
+
+        // act
+        for (int i = 1; i < MAX_FILES_TO_UPLOAD; i++)
         {
-            fileUploadNotification = fileUploadNotificationReceiver.receive(MAXIMUM_TIME_TO_WAIT_FOR_IOTHUB);
-            assertNotNull(fileUploadNotification);
+            final int index = i;
+            executor.submit(new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    try
+                    {
+                        deviceClient.uploadToBlobAsync(fileUploadState[index].blobName, fileUploadState[index].fileInputStream, fileUploadState[index].fileLength, new FileUploadCallback(), fileUploadState[index]);
+                    }
+                    catch (IOException e)
+                    {
+                        fail(e.getMessage());
+                    }
+                }
+            });
 
-            //ignore any file upload notifications received that are not about this device
-        } while (!fileUploadNotification.getDeviceId().equals(scDevice.getDeviceId()));
+            executor.submit(new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    deviceClient.sendEventAsync(new com.microsoft.azure.sdk.iot.device.Message(messageStates[index].messageBody), new FileUploadCallback(), messageStates[index]);
+                }
+            });
 
-        return fileUploadNotification;
+            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice(scDevice);
+
+            // assert
+            verifyNotification(fileUploadNotification, fileUploadState[i]);
+
+            assertTrue(fileUploadState[i].isCallBackTriggered);
+            assertEquals(fileUploadState[i].fileUploadStatus, SUCCESS);
+            assertEquals(messageStates[i].messageStatus, SUCCESS);
+        }
+
+        executor.shutdown();
+        if (!executor.awaitTermination(10000, TimeUnit.MILLISECONDS))
+        {
+            executor.shutdownNow();
+        }
+
+        for (int i = 1; i < MAX_FILES_TO_UPLOAD; i++)
+        {
+            assertEquals("File" + i + " has no notification", fileUploadState[i].fileUploadNotificationReceived, SUCCESS);
+        }
+
+        tearDownDeviceClient();
     }
 
     @Test (timeout = MAX_MILLISECS_TIMEOUT_KILL_TEST)
-    public void uploadToBlobAsyncSingleFileAndTelemetryOnAMQP() throws URISyntaxException, IOException, InterruptedException, ExecutionException, TimeoutException
+    public void uploadToBlobAsyncSingleFileOnAMQP() throws URISyntaxException, IOException, InterruptedException, ExecutionException, TimeoutException
     {
         // arrange
         setUpDeviceClient(IotHubClientProtocol.AMQPS);
@@ -460,7 +525,7 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
                 }
             });
 
-            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice();
+            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice(scDevice);
 
             // assert
             verifyNotification(fileUploadNotification, fileUploadState[i]);
@@ -484,7 +549,7 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
     }
 
     @Test (timeout = MAX_MILLISECS_TIMEOUT_KILL_TEST)
-    public void uploadToBlobAsyncSingleFileAndTelemetryOnAMQPWS() throws URISyntaxException, IOException, InterruptedException, ExecutionException, TimeoutException
+    public void uploadToBlobAsyncSingleFileOnAMQPWS() throws URISyntaxException, IOException, InterruptedException, ExecutionException, TimeoutException
     {
         // arrange
         setUpDeviceClient(IotHubClientProtocol.AMQPS_WS);
@@ -519,7 +584,7 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
                 }
             });
 
-            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice();
+            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice(scDevice);
 
             // assert
             verifyNotification(fileUploadNotification, fileUploadState[i]);
@@ -543,7 +608,7 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
     }
 
     @Test (timeout = MAX_MILLISECS_TIMEOUT_KILL_TEST)
-    public void uploadToBlobAsyncSingleFileAndTelemetryOnHttp() throws URISyntaxException, IOException, InterruptedException, ExecutionException, TimeoutException
+    public void uploadToBlobAsyncSingleFileOnHttp() throws URISyntaxException, IOException, InterruptedException, ExecutionException, TimeoutException
     {
         // arrange
         setUpDeviceClient(IotHubClientProtocol.HTTPS);
@@ -578,7 +643,7 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
                 }
             });
 
-            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice();
+            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice(scDevice);
 
             // assert
             verifyNotification(fileUploadNotification, fileUploadState[i]);
@@ -599,5 +664,197 @@ public class FileUploadTests extends MethodNameLoggingIntegrationTest
         }
 
         tearDownDeviceClient();
+    }
+
+    @Test (timeout = MAX_MILLISECS_TIMEOUT_KILL_TEST)
+    public void uploadToBlobAsyncSingleFileOnHttpSelfSigned() throws URISyntaxException, IOException, InterruptedException, ExecutionException, TimeoutException
+    {
+        // arrange
+        setUpX509DeviceClient(IotHubClientProtocol.HTTPS);
+        ExecutorService executor = Executors.newFixedThreadPool(2);
+
+        // act
+        for (int i = 1; i < MAX_FILES_TO_UPLOAD; i++)
+        {
+            final int index = i;
+            executor.submit(new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    try
+                    {
+                        deviceClient.uploadToBlobAsync(fileUploadState[index].blobName, fileUploadState[index].fileInputStream, fileUploadState[index].fileLength, new FileUploadCallback(), fileUploadState[index]);
+                    }
+                    catch (IOException e)
+                    {
+                        fail(e.getMessage());
+                    }
+                }
+            });
+
+            executor.submit(new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    deviceClient.sendEventAsync(new com.microsoft.azure.sdk.iot.device.Message(messageStates[index].messageBody), new FileUploadCallback(), messageStates[index]);
+                }
+            });
+
+            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice(scDevicex509);
+
+            // assert
+            verifyNotification(fileUploadNotification, fileUploadState[i]);
+            assertTrue(fileUploadState[i].isCallBackTriggered);
+            assertEquals(fileUploadState[i].fileUploadStatus, SUCCESS);
+            assertEquals(messageStates[i].messageStatus, SUCCESS);
+        }
+
+        executor.shutdown();
+        if (!executor.awaitTermination(10000, TimeUnit.MILLISECONDS))
+        {
+            executor.shutdownNow();
+        }
+
+        for (int i = 1; i < MAX_FILES_TO_UPLOAD; i++)
+        {
+            assertEquals("File" + i + " has no notification", fileUploadState[i].fileUploadNotificationReceived, SUCCESS);
+        }
+
+        tearDownDeviceClient();
+    }
+
+    @Test (timeout = MAX_MILLISECS_TIMEOUT_KILL_TEST)
+    public void uploadToBlobAsyncAndTelemetryOnMQTTSelfSigned() throws URISyntaxException, IOException, InterruptedException, ExecutionException, TimeoutException
+    {
+        // arrange
+        setUpX509DeviceClient(IotHubClientProtocol.MQTT);
+        ExecutorService executor = Executors.newFixedThreadPool(2);
+
+        // act
+        for (int i = 1; i < MAX_FILES_TO_UPLOAD; i++)
+        {
+            final int index = i;
+            executor.submit(new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    try
+                    {
+                        deviceClient.uploadToBlobAsync(fileUploadState[index].blobName, fileUploadState[index].fileInputStream, fileUploadState[index].fileLength, new FileUploadCallback(), fileUploadState[index]);
+                    }
+                    catch (IOException e)
+                    {
+                        fail(e.getMessage());
+                    }
+                }
+            });
+
+            executor.submit(new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    deviceClient.sendEventAsync(new com.microsoft.azure.sdk.iot.device.Message(messageStates[index].messageBody), new FileUploadCallback(), messageStates[index]);
+                }
+            });
+
+            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice(scDevicex509);
+
+            // assert
+            verifyNotification(fileUploadNotification, fileUploadState[i]);
+
+            assertTrue(fileUploadState[i].isCallBackTriggered);
+            assertEquals(fileUploadState[i].fileUploadStatus, SUCCESS);
+            assertEquals(messageStates[i].messageStatus, SUCCESS);
+        }
+
+        executor.shutdown();
+        if (!executor.awaitTermination(10000, TimeUnit.MILLISECONDS))
+        {
+            executor.shutdownNow();
+        }
+
+        for (int i = 1; i < MAX_FILES_TO_UPLOAD; i++)
+        {
+            assertEquals("File" + i + " has no notification", fileUploadState[i].fileUploadNotificationReceived, SUCCESS);
+        }
+
+        tearDownDeviceClient();
+    }
+
+    @Test (timeout = MAX_MILLISECS_TIMEOUT_KILL_TEST)
+    public void uploadToBlobAsyncSingleFileOnAMQPSelfSigned() throws URISyntaxException, IOException, InterruptedException, ExecutionException, TimeoutException
+    {
+        // arrange
+        setUpX509DeviceClient(IotHubClientProtocol.AMQPS);
+        ExecutorService executor = Executors.newFixedThreadPool(2);
+
+        // act
+        for (int i = 1; i < MAX_FILES_TO_UPLOAD; i++)
+        {
+            final int index = i;
+            executor.submit(new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    try
+                    {
+                        deviceClient.uploadToBlobAsync(fileUploadState[index].blobName, fileUploadState[index].fileInputStream, fileUploadState[index].fileLength, new FileUploadCallback(), fileUploadState[index]);
+                    }
+                    catch (IOException e)
+                    {
+                        fail(e.getMessage());
+                    }
+                }
+            });
+
+            executor.submit(new Runnable()
+            {
+                @Override
+                public void run()
+                {
+                    deviceClient.sendEventAsync(new com.microsoft.azure.sdk.iot.device.Message(messageStates[index].messageBody), new FileUploadCallback(), messageStates[index]);
+                }
+            });
+
+            FileUploadNotification fileUploadNotification = getFileUploadNotificationForThisDevice(scDevicex509);
+
+            // assert
+            verifyNotification(fileUploadNotification, fileUploadState[i]);
+            assertTrue(fileUploadState[i].isCallBackTriggered);
+            assertEquals(fileUploadState[i].fileUploadStatus, SUCCESS);
+            assertEquals(messageStates[i].messageStatus, SUCCESS);
+        }
+
+        executor.shutdown();
+        if (!executor.awaitTermination(10000, TimeUnit.MILLISECONDS))
+        {
+            executor.shutdownNow();
+        }
+
+        for (int i = 1; i < MAX_FILES_TO_UPLOAD; i++)
+        {
+            assertEquals("File" + i + " has no notification", fileUploadState[i].fileUploadNotificationReceived, SUCCESS);
+        }
+
+        tearDownDeviceClient();
+    }
+
+    private FileUploadNotification getFileUploadNotificationForThisDevice(Device device) throws IOException, InterruptedException
+    {
+        FileUploadNotification fileUploadNotification;
+        do
+        {
+            fileUploadNotification = fileUploadNotificationReceiver.receive(MAXIMUM_TIME_TO_WAIT_FOR_IOTHUB);
+            assertNotNull(fileUploadNotification);
+
+            //ignore any file upload notifications received that are not about this device
+        } while (!fileUploadNotification.getDeviceId().equals(device.getDeviceId()));
+
+        return fileUploadNotification;
     }
 }

--- a/iot-e2e-tests/jvm/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/iothubservices/FileUploadJVMRunner.java
+++ b/iot-e2e-tests/jvm/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/iothubservices/FileUploadJVMRunner.java
@@ -7,17 +7,25 @@ package tests.integration.com.microsoft.azure.sdk.iot.iothubservices;
 
 import com.microsoft.azure.sdk.iot.common.helpers.TestConstants;
 import com.microsoft.azure.sdk.iot.common.helpers.Tools;
+import com.microsoft.azure.sdk.iot.common.helpers.X509Cert;
 import com.microsoft.azure.sdk.iot.common.tests.iothubservices.FileUploadTests;
 import org.junit.BeforeClass;
 
 import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateEncodingException;
 
 public class FileUploadJVMRunner extends FileUploadTests
 {
     @BeforeClass
-    public static void setup() throws IOException
+    public static void setup() throws IOException, GeneralSecurityException
     {
         iotHubConnectionString = Tools.retrieveEnvironmentVariableValue(TestConstants.IOT_HUB_CONNECTION_STRING_ENV_VAR_NAME);
-        FileUploadTests.setUp();
+        X509Cert cert = new X509Cert(0,false, "TestLeaf", "TestRoot");
+        String publicKeyCert = cert.getPublicCertLeafPem();
+        String privateKey =  cert.getPrivateKeyLeafPem();
+        String x509Thumbprint = cert.getThumbPrintLeaf();
+        FileUploadTests.setUp(publicKeyCert, privateKey, x509Thumbprint);
     }
 }

